File: app/src/main/java/com/akhbulatov/githubsample/ui/global/BasePresenterImpl.java
package com.akhbulatov.githubsample.ui.global;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.HashSet;
import java.util.Set;

import retrofit2.Call;

public abstract class BasePresenterImpl<T extends BaseView> implements BasePresenter<T> {

    @Nullable private T view;
    @NonNull private Set<Call<?>> requestSet = new HashSet<>();

    @Override public void attachView(@NonNull T view) {
        this.view = view;
    }

    @Override public void detachView() {
        cancelRequests();
        view = null;
    }

    @Nullable protected T getView() {
        return view;
    }

    protected void addRequest(Call<?> request) {
        requestSet.add(request);
    }

    private void cancelRequests() {
        for (Call<?> request : requestSet) {
            request.cancel();
        }
    }
}

File: app/src/main/java/com/akhbulatov/archsample/ui/global/BasePresenterImpl.kt
package com.akhbulatov.archsample.ui.global

import retrofit2.Call
import java.util.*

abstract class BasePresenterImpl<T : BaseView> : BasePresenter<T> {

    protected var view: T? = null
    private val requestSet = HashSet<Call<*>>()

    override fun attachView(view: T) {
        this.view = view
    }

    override fun detachView() {
        cancelRequests()
        view = null
    }

    protected fun addRequest(request: Call<*>) {
        requestSet.add(request)
    }

    private fun cancelRequests() {
        for (request in requestSet) {
            request.cancel()
        }
    }
}